pkg/controllers/provisioner/suite_test.go

```diff
@@ -0,0 +1,125 @@
+limitations under the License.
+*/
+
+package provisioner_test
+
+import (
+	"context"
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/samber/lo"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	. "knative.dev/pkg/logging/testing"
+	"knative.dev/pkg/ptr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/aws/karpenter-core/pkg/apis"
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	provcontroller "github.com/aws/karpenter-core/pkg/controllers/provisioner"
+	"github.com/aws/karpenter-core/pkg/operator/controller"
+	"github.com/aws/karpenter-core/pkg/operator/scheme"
+	"github.com/aws/karpenter-core/pkg/test"
+	. "github.com/aws/karpenter-core/pkg/test/expectations"
+)
+
+var provisionerController controller.Controller
+var provisioner *v1alpha5.Provisioner
+var ctx context.Context
+var env *test.Environment
+
+func TestAPIs(t *testing.T) {
+	ctx = TestContextWithLogger(t)
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Controllers/Provisioner")
+}
+
+var _ = BeforeSuite(func() {
+	env = test.NewEnvironment(scheme.Scheme, test.WithCRDs(apis.CRDs...))
+	provisionerController = provcontroller.NewController(env.Client)
+	provisioner = test.Provisioner(test.ProvisionerOptions{
+		Taints: []v1.Taint{
+			{
+				Key:    "key",
+				Effect: v1.TaintEffectNoExecute,
+			},
+		},
+		StartupTaints: []v1.Taint{
+			{
+				Key:    "key",
+				Effect: v1.TaintEffectNoExecute,
+			},
+		},
+		Labels: map[string]string{
+			"keyLabel": "valueLabel",
+		},
+		Kubelet: &v1alpha5.KubeletConfiguration{
+			MaxPods: ptr.Int32(10),
+		},
+		Annotations: map[string]string{
+			"keyAnnotation":  "valueAnnotation",
+			"keyAnnotation2": "valueAnnotation2",
+		},
+	})
+})
+
+var _ = AfterSuite(func() {
+	Expect(env.Stop()).To(Succeed(), "Failed to stop environment")
+})
+
+var _ = Describe("Provisioner Static Drift Hash", func() {
+	It("should update the static drift hash when provisioner static field is updated", func() {
+		ExpectApplied(ctx, env.Client, provisioner)
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+
+		expectedHash := provisioner.Hash()
+		Expect(provisioner.ObjectMeta.Annotations[v1alpha5.ProvisionerHashAnnotationKey]).To(Equal(expectedHash))
+
+		provisioner.Spec.Labels = map[string]string{"keyLabeltest": "valueLabeltest"}
+		provisioner.Spec.Annotations = map[string]string{"keyAnnotation2": "valueAnnotation2", "keyAnnotation": "valueAnnotation"}
+		ExpectReconcileSucceeded(ctx, provisionerController, client.ObjectKeyFromObject(provisioner))
+		provisioner = ExpectExists(ctx, env.Client, provisioner)
+
+		expectedHashTwo := provisioner.Hash()
+		Expect(provisioner.ObjectMeta.Annotations[v1alpha5.ProvisionerHashAnnotationKey]).To(Equal(expectedHashTwo))
+	})
+	It("should not update the static drift hash when provisioner behavior field is updated", func() {
```

Comment:
Just a note: On the other hand, since we're expecting all of these fields to not have an affect on the hash, changing all behavioral fields and checking no hash change would be sufficient here.